knowledgebase/importing-geojason-with-nested-object-array.md

```diff
@@ -0,0 +1,222 @@
+```
+Here, we get the following error:
+>Received exception from server (version 24.1.2):
+Code: 53. DB::Exception: Received from localhost:9000. DB::Exception: ARRAY JOIN requires array or map argument. (TYPE_MISMATCH)
+
+This is caused by the parsing of `features.geometry.coordinates`.
+
+6. Let's check its data type.
+
+``` sql
+SELECT DISTINCT toTypeName(features.geometry.coordinates) AS geometry
+FROM file('municipios_ign.geojson', 'JSON')
+ARRAY JOIN features
+
+┌─geometry──────────────────────────────────────┐
+│ Array(Array(Array(Array(Nullable(Float64))))) │
+└───────────────────────────────────────────────┘
+```
+
+It can be fixed by casting `multipolygon.properties.coordinates` to `Array(Array(Array(Tuple(Float64,Float64))))`.
+To do so, we can use the function [arrayMap(func,arr1,...)](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#arraymapfunc-arr1-).
+
+```sql
+SELECT distinct
+    toTypeName(
+        arrayMap(features.geometry.coordinates->
+                    arrayMap(features.geometry.coordinates-> 
+                            arrayMap(features.geometry.coordinates-> (features.geometry.coordinates[1],features.geometry.coordinates[2]) 
+                    ,features.geometry.coordinates),
+                features.geometry.coordinates),
+        features.geometry.coordinates)
+    ) as toTypeName
+FROM file('municipios_ign.geojson', 'JSON')
+ARRAY JOIN features;
+
+┌─toTypeName───────────────────────────────────────────────────────┐
+│ Array(Array(Array(Tuple(Nullable(Float64), Nullable(Float64))))) │
+└──────────────────────────────────────────────────────────────────┘
+```
+
+7. Insert the data.
+
+```sql
+INSERT INTO geojson
+SELECT
+    type as type,
+    name as name,
+    crs.type as crsType,
+    crs.properties.name as crsName,
+    features.type as featureType,
+    features.properties.FID id,
+    features.properties.INSPIREID inspiredId,
+    features.properties.NATCODE natCode,
+    features.properties.NAMEUNIT nameUnit,
+    features.properties.CODNUT1 codNut1,
+    features.properties.CODNUT2 codNut2,
+    features.properties.CODNUT3 codNut3,
+    features.properties.CODIGOINE codigoIne,
+    features.properties.SHAPE_Length shapeLength,
+    features.properties.SHAPE_Area shapeArea,
+    features.geometry.type geometryType,
+    arrayMap(features.geometry.coordinates->
+                arrayMap(features.geometry.coordinates-> 
+                        arrayMap(features.geometry.coordinates-> (features.geometry.coordinates[1],features.geometry.coordinates[2]),features.geometry.coordinates)
+                ,features.geometry.coordinates)
+    ,features.geometry.coordinates) geometry
+FROM file('municipios_ign.geojson', 'JSON')
+ARRAY JOIN features;
+```
+
+```sql
+SELECT count()
+FROM geojson
+
+┌─count()─┐
+│    8205 │
+└─────────┘
+
+SELECT DISTINCT toTypeName(geometry)
+FROM geojson
+
+┌─toTypeName(geometry)─┐
+│ MultiPolygon         │
+└──────────────────────┘
+```
+
```

Comment:
In addition to the `count()` query, it may be good to demo something complex to show the analytical power of geo data in ClickHouse. Perhaps something that filters on one of the many columns? We also have some functions for measuring distance, but I don't think they can be easily applied to the polygon data

Reply:
I'll look into: 
- https://clickhouse.com/docs/en/sql-reference/functions/geo/polygons#polygonareaspherical
- https://clickhouse.com/docs/en/sql-reference/functions/geo/polygons#polygonperimeterspherical
- https://clickhouse.com/docs/en/sql-reference/functions/geo/polygons#polygonareacartesian
- https://clickhouse.com/docs/en/sql-reference/functions/geo/polygons#polygonperimetercartesian